ddl/ddl_api.go

```diff
@@ -566,10 +566,25 @@ func buildColumnAndConstraint(
 	if err != nil {
 		return nil, nil, errors.Trace(err)
 	}
-
 	if err := setCharsetCollationFlenDecimal(colDef.Tp, chs, coll); err != nil {
 		return nil, nil, errors.Trace(err)
 	}
+	// Some column checks are performed before convert, so varchar auto convert is checked here, see issue #30328
```

Comment:
Why check and revert it here? Is it possible to only check and convert in `preprocessor.checkCreateTableGrammar`?

Reply:
Because in some situations charset of the column is not resolved in `preprocessor.checkCreateTableGrammar`. For example `CREATE TABLE t1(a varbinary(70000), b varchar(70000000));`, the charset and collation of `b` is resolved in func `buildColumnAndConstraint`.

Reply:
Then is it possible to only add the code in `buildColumnAndConstraint`?

Reply:
IMHO it is not possible to do so. `buildColumnAndConstraint` builds `table.Column` from `ast.ColumnDef`, and it is called from `(*ddl).CreateTable`, while `(*ddl).getModifiableColumnJob` builds `model.Job` from `ast.AlterTableSpec`, which is called in `(*ddl).ChangeColumn` and `(*ddl).ModifyColumn`.